### PR TITLE
core: Defer script extraction until we know it's needed

### PR DESCRIPTION
**Pull request by cdecker**
We were always allocating space and copying the scripts from the
outputs, and then immediately discarding them again if they weren't
P2WSH outputs. This patch adds an introspection method, and then uses
it to determine whether we're interested in the output before cloning
the script. This should save us almost 1 allocation and 1 deallocation
for each output in a block (thousands).

Builds on top of #6983

### CODE REVIEW

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/6984/commits/efca2991c3af471745b0b5acafb2df503046dc39

Review the last two commits! and reviewed the https://github.com/ElementsProject/lightning/pull/6983 as well

I do not know if you are finishing to work on this or it is just a draft because the CI is failing.

Regarding the CI looks like there is just a server flakyness

```
Secret source: Actions
Prepare workflow directory
Prepare all required actions
Getting action download info
Error: Internal server error occurred while resolving "actions/checkout@v3". Internal server error occurred while resolving "actions/download-artifact@v3". Internal server error occurred while resolving "actions/setup-python@v4"
```

File: bitcoin/tx.c [hunk 1]
```diff
@@ -324,6 +325,16 @@ const u8 *bitcoin_tx_output_get_script(const tal_t *ctx,
...
 }
 
+bool bitcoin_tx_output_script_is_p2wsh(const struct bitcoin_tx *tx, int outnum)
+{	const struct wally_tx_output *output;
+	assert(outnum < tx->wtx->num_outputs);
+	output = &tx->wtx->outputs[outnum];
+
+	return output->script_len == BITCOIN_SCRIPTPUBKEY_P2WSH_LEN &&
+	       output->script[0] == OP_0 &&
+	       output->script[1] != OP_PUSHBYTES(sizeof(struct ripemd160));
```
Comment (jgriffiths) [inline bitcoin/tx.c:335]: This should be the sizeof a sha256 struct, not ripemd160 I believe:

- `OP_0 [20 byte hash160]` -> p2wpkh
- `OP_O [32 byte sha256]` -> p2wsh

File: lightningd/chaintopology.c [hunk 2]
```diff
@@ -955,9 +957,11 @@ static void topo_add_utxos(struct chain_topology *topo, struct block *b)
 			if (!amount_asset_is_main(&amt))
 				continue;
 
-			const u8 *script = bitcoin_tx_output_get_script(tmpctx, tx, n);
-			if (is_p2wsh(script, NULL)) {
+			if (bitcoin_tx_output_script_is_p2wsh(tx, n)) {
+				outpoint.txid = b->txids[i];
```
Comment (jgriffiths) [inline lightningd/chaintopology.c:961]: If you are going to re-copy the txid, you should remove it from the init of `struct bitcoin_outpoint outpoint` above, at which point you might as well move the struct definition here and init it inline.

**Comment by jgriffiths** [comment]
Also worth noting that all Elements/Liquid txs will have a fee output which passes the policy asset check and fails the script check, which you can skip by ignoring outputs with a zero `script_len` before the asset check.

**Comment by jgriffiths** [comment]
One final note, I believe that this code in `get_new_block`:

```
        /* Annotate all transactions with the chainparams */
        for (size_t i = 0; i < tal_count(blk->tx); i++)
                blk->tx[i]->chainparams = chainparams;
```

Is redundant because its already set in `bitcoin_block_from_hex`:

```
       for (i = 0; i < num; i++) {
                ...
                b->tx[i]->chainparams = chainparams;
                ...
        }
```

**Comment by cdecker** [comment]
Thanks @jgriffiths that is very helpful. I spent the night running benchmarks, and for some reason the performance of both of our patches ends up _worse_ than just yours, despite allocating less (???). It could very well be that I mistakenly picked `ripemd160` and therefore end up doing the majority of the allocations again, but I'm still looking :+1:

**Comment by jgriffiths** [comment]
@cdecker The incorrect push size check means that the utxos of interest won't be being tracked so its an apples to oranges comparison.

Thinking more about this code though, and given that fee outputs likely make up somewhere between 30-50% of Elements outputs, an overall faster and simpler approach is probably just to early-out based on the script length, and avoid adding any new introspection function. We can also skip pegin outputs for free when checking for coinbase txs, and remove the `tal_count` overhead as described elsewhere.

I've added the above changes to https://github.com/ElementsProject/lightning/pull/6983 and rebased it, so you can directly benchmark that PR against master.

File: bitcoin/tx.c (hunk 1 above)
Comment (cdecker) [reply bitcoin/tx.c:335]: Thanks @jgriffiths, this actually got the performance up again.

File: lightningd/chaintopology.c (hunk 2 above)
Comment (cdecker) [reply lightningd/chaintopology.c:961]: As far as I can see the txid is copied only once, and the declaration at the top of the function reserves room, but doesn't initialize anymore.

**Comment by cdecker** [comment]
I finally got around to benchmarking the changes in #6983 and this PR. And the results are quite nice, thanks to @jgriffiths. The benchmarks were run 5 times each, on the same range of blocks from 824000 to 825000, cached by a reverse proxy to minimize time we spend asking `bitcoind` for block data, and we pick the median run as representative:

| Name          | Blocks | Time [s] | Blocks / s | Speedup [%] |
|---------------|--------|----------|------------|------------:|
| Baseline      | 1000   | 193      | 5.164      |          0% |
| #6983         | 1000   | 151      | 6.611      |       28.0% |
| #6983 & #6984 | 1000   | 127      | 7.818      |       51.3% |

This is still with the older version of #6983, that did not cache the `tal_count()` calls and skip empty scripts, so we might even get out a bit more perf, but I think these results are sufficient to show their usefulness :-)

**Comment by cdecker** [comment]
Converted back to Draft, since we appear to accidentally break some tests.

File: bitcoin/tx.c [hunk 3]
```diff
@@ -324,6 +324,16 @@ const u8 *bitcoin_tx_output_get_script(const tal_t *ctx,
...
 }
 
+bool bitcoin_tx_output_script_is_p2wsh(const struct bitcoin_tx *tx, int outnum)
+{	const struct wally_tx_output *output;
+	assert(outnum < tx->wtx->num_outputs);
+	output = &tx->wtx->outputs[outnum];
+
+	return output->script_len == BITCOIN_SCRIPTPUBKEY_P2WSH_LEN &&
+	       output->script[0] == OP_0 &&
+	       output->script[1] != OP_PUSHBYTES(sizeof(struct sha256));
```
[the remark this answers is not in the document]
Comment (jgriffiths) [reply bitcoin/tx.c:334]: Note that `process_getfilteredblock_step1` has the same non-optimal script copying behaviour. In fact, every call to `bitcoin_tx_output_get_script` in the code is redundant. Take a look at `calc_tx_fee`, for example:

```
                oscript = bitcoin_tx_output_get_script(NULL, tx, i);
                scriptlen = tal_bytelen(oscript);
                tal_free(oscript);

                if (chainparams->is_elements && scriptlen == 0)
                        continue;
```

This should just be:
```
                if (chainparams->is_elements && !tx->wtx->outputs[i].script_len)
                        continue; /* Fee output, ignore */
```

(Ignoring that the code here ignores the fee, and should probably reconcile against it).

**Pull request merged**